lib/gitsh/git_command_list.rb

```diff
@@ -5,7 +5,38 @@ def initialize(env)
     end
 
     def to_a
-      git_output('help -a').
+      commands_from_list_cmds ||
+        commands_from_help('help -a --no-verbose') ||
+        commands_from_help('help -a') ||
+        []
+    end
+
+    private
+
+    attr_accessor :env
+
+    def commands_from_list_cmds
+      output = git_output('--list-cmds=main,nohelpers')
+
+      if output
+        output.lines.map(&:chomp).sort
+      else
+        false
```

Comment:
How about returning `[]` in the failure case, and then instead of `||` you make a helper method that calls the next result if the result is `empty?`?

Reply:
To make that lazy (which I think it should be, because each of these runs an external command) it would probably need to use some nested blocks. I guess that would be fine though. Something like:

```ruby
try_using(commands_from_list_cmds) do
  try_using(commands_from_help('help -a --no-verbose')) do
    try_using(commands_from_help('help -a'), default: [])
  end
end
```

Reply:
Yeah.

But on the up-side, this method gets simpler:

```ruby
git_output('--list-cmds=main,nohelpers')&.lines.map(&:chomp).sort || []
```

though maybe it doesn't help the other two...

Reply:
I don’t know which version of Ruby you’re targeting, but would some combination of `&.` and `yield_self` be appropriate?

```ruby
def commands_from_list_cmds
  git_output('--list-cmds=main,nohelpers')&.
    lines.
    map(&:chomp).
    sort
end

def commands_from_help(command)
  git_output(command)&.yield_self(:parse_help_output)
end
```

Reply:
Currently, we support Ruby 2.0 and up (https://github.com/thoughtbot/gitsh/blob/master/configure.ac#L30-L34) but since 2.0–2.2 are now EOL'd, I'd be open to requiring 2.3.

`&.` was introduced in 2.3, but `#yield_self` is probably still out of reach in 2.5.

Reply:
I've tried going down the `#try_using` route, combined with making sure `#git_output` always returns an array of lines (which could be empty). I think the result is much better—what do y'all think?